File: code/demonstration/receiver/AndroidIteration3/app/src/main/java/de/fzi/sensidl/receiver/android/iteration3/MainActivity.java
package de.fzi.sensidl.receiver.android.iteration3;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.bluetooth.le.BluetoothLeScanner;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanResult;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements CompoundButton.OnCheckedChangeListener, View.OnClickListener {

	public static final String DEVICE_NAME    = "deviceName";
	public static final String DEVICE_ADDRESS = "deviceAddress";

	private Switch  searchForDevices;
	private Spinner nearDevices;
	private Button  connectToDevice;

	private BluetoothLeScanner   bluetoothLeScanner;
	private BluetoothAdapter     bluetoothAdapter;
	private ArrayAdapter<String> spinnerAdapter;
	private List<BluetoothDevice> bluetoothList = new ArrayList<>();

	/**
	 * A ScanCallback object, that will be called when there is a new bluetooth device near
	 */
	private ScanCallback leScanCallback = new ScanCallback() {
		@Override
		public void onScanResult(int callbackType, ScanResult result) {
			if (!bluetoothList.contains(result.getDevice())) {
				addDeviceToBluetoothList(result.getDevice());
			}
		}
	};

	/**
	 * The onCreate method called to create the view Find the views and set an on click listener on the button
	 *
	 * @param savedInstanceState
	 */
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		searchForDevices = (Switch) findViewById(R.id.search_for_sensors);
		nearDevices = (Spinner) findViewById(R.id.near_sensors);
		connectToDevice = (Button) findViewById(R.id.connect_to_sensor);

		searchForDevices.setOnCheckedChangeListener(this);
		connectToDevice.setOnClickListener(this);

		spinnerAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, android.R.id.text1);
		spinnerAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
		nearDevices.setAdapter(spinnerAdapter);
		initEmptyList();
		isBLEAvailable();
		getBluetoothLeScanner();
	}

	/**
	 * Will be called when the activity stops
	 */
	@Override
	protected void onStop() {
		stopBluetoothLeScanner();
		super.onStop();
	}

	/**
	 * Get the bluetooth le scanner from the system
	 */
	private void getBluetoothLeScanner() {
		bluetoothAdapter = ((BluetoothManager) getSystemService(Context.BLUETOOTH_SERVICE)).getAdapter();
		bluetoothLeScanner = bluetoothAdapter.getBluetoothLeScanner();
	}

	/**
	 * Stop the bluetooth le scanner
	 */
	private void stopBluetoothLeScanner() {
		searchForDevices.setChecked(false);
		bluetoothAdapter = null;
		bluetoothLeScanner = null;
	}

	/**
	 * Initialize an empty list
	 */
	private void initEmptyList() {
		spinnerAdapter.clear();
		spinnerAdapter.add(getString(R.string.no_sensor_selected));
		spinnerAdapter.notifyDataSetChanged();
	}

	/**
	 * Will be called when a switch changes his status
	 *
	 * @param compoundButton the button
	 * @param b              if the switch is set to true or false
	 */
	@Override
	public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
		if (compoundButton.getId() == searchForDevices.getId()) {
			getBluetoothLeScanner();
			if (bluetoothLeScanner != null) {
				scanForSensors(b);
			} else {
				Toast.makeText(this, R.string.bluetooth_not_activated, Toast.LENGTH_LONG).show();
				compoundButton.setChecked(false);
			}
		}
	}

	/**
	 * Activate scanning for near devices
	 *
	 * @param scan
	 */
	private void scanForSensors(boolean scan) {
		if (scan) {
			clearBluetoothDeviceLists();
			bluetoothLeScanner.startScan(leScanCallback);
		} else {
			bluetoothLeScanner.stopScan(leScanCallback);
			searchForDevices.setChecked(false);
		}
	}

	/**
	 * Check if bluetooth le is available on the system. If not close the app.
	 */
	private void isBLEAvailable() {
		if (!getPackageManager().hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE)) {
			Toast.makeText(this, R.string.no_ble_available, Toast.LENGTH_LONG).show();
			finish();
		}
	}

	/**
	 * Ad a near device to the list
	 *
	 * @param bluetoothDevice
	 */
	protected void addDeviceToBluetoothList(BluetoothDevice bluetoothDevice) {
		if (bluetoothDevice.getName() != null) {
			bluetoothList.add(bluetoothDevice);
			spinnerAdapter.add(bluetoothDevice.getName());
			spinnerAdapter.notifyDataSetChanged();
		}
	}

	/**
	 * clear the lists that holding the bluetooth devices
	 */
	private void clearBluetoothDeviceLists() {
		bluetoothList.clear();
		initEmptyList();
	}

	/**
	 * When a view get clicked
	 *
	 * @param view
	 */
	@Override
	public void onClick(View view) {
		if (view.getId() == connectToDevice.getId()) {
			onConnectToDeviceClicked();
		}
	}

	/**
	 * Connect to the selected device
	 */
	private void onConnectToDeviceClicked() {
		if (nearDevices.getSelectedItemPosition() != 0) {
			BluetoothDevice selectedDevice = bluetoothList.get(nearDevices.getSelectedItemPosition() - 1);
			scanForSensors(false);
			Intent intent = new Intent(this, DeviceControlActivity.class);
			intent.putExtra(DEVICE_NAME, selectedDevice.getName());
			intent.putExtra(DEVICE_ADDRESS, selectedDevice.getAddress());
			startActivity(intent);
		} else {
			Toast.makeText(this, R.string.no_sensor_selected, Toast.LENGTH_LONG).show();
		}
	}
}

File: code/demonstration/receiver/XamarinIteration3/XamarinStufe3/XamarinStufe3.Droid/MainActivity.cs
using System;

using Android.App;
using Android.Views;
using Android.Widget;
using Android.OS;
using Android.Support.V7.App;
using Robotics.Mobile.Core.Bluetooth.LE;
using System.Collections.Generic;
using Android.Content.PM;

namespace XamarinIteration3.Droid
{
	[Activity (Label = "@string/app_name", MainLauncher = true, Icon = "@drawable/icon", ScreenOrientation = ScreenOrientation.Portrait)]
	public class MainActivity : AppCompatActivity, View.IOnClickListener, CompoundButton.IOnCheckedChangeListener
	{

		Spinner _nearDevices;
		Switch _searchForDevices;
		Button _connectToDevice;
		MyDeviceService _deviceService;
		List<Device> _devices;
		ArrayAdapter<String> _spinnerAdapter;

		// <summary>
		// The onCreate method called to create the view Find the views and set an on click listener on the button
		// </summary>
		// <param name="savedInstanceState">savedInstanceState</param>
		protected override void OnCreate (Bundle savedInstanceState)
		{
			base.OnCreate (savedInstanceState);
			SetContentView (Resource.Layout.main_activity);
			_deviceService = MyDeviceService.Instance (new Robotics.Mobile.Core.Bluetooth.LE.Adapter ());
			_devices = new List<Device> ();

			_connectToDevice = FindViewById<Button> (Resource.Id.connect_to_device);
			_connectToDevice.SetOnClickListener (this);
			_searchForDevices = FindViewById<Switch> (Resource.Id.search_for_devices);
			_searchForDevices.SetOnCheckedChangeListener (this);

			_nearDevices = FindViewById<Spinner> (Resource.Id.near_devices);
			_spinnerAdapter = new ArrayAdapter<string> (this, Android.Resource.Layout.SimpleSpinnerItem, Android.Resource.Id.Text1);
			_spinnerAdapter.SetDropDownViewResource (Android.Resource.Layout.SimpleSpinnerDropDownItem);
			_nearDevices.Adapter = _spinnerAdapter;
			InitEmptyLists ();
		}

		// <summary>
		// Called when the activity goes on pause
		// </summary>
		protected override void OnPause ()
		{
			ScanningForDevices (false);
			base.OnPause ();
		}

		// <summary>
		// Start scanning for devices
		// </summary>
		// <param name="scan">If we scan (true) or not (false)</param>
		void ScanningForDevices (Boolean scan)
		{
			InitEmptyLists ();
			_deviceService.ScanningForDevices (scan, AddDeviceToList);
		}

		// <summary>
		// Check if ble is available on the device
		// </summary>
		Boolean IsBLEAvailable ()
		{
			if (!PackageManager.HasSystemFeature (Android.Content.PM.PackageManager.FeatureBluetoothLe)) {
				Toast.MakeText (this, "Kein BLE verfügbar", ToastLength.Long).Show ();
				Finish ();
				return false;
			} else {
				return true;
			}
		}

		// <summary>
		// Add a device in the near to the list
		// </summary>
		// <param name="iDevice">the iDevice we add to the list</param>
		public void AddDeviceToList (IDevice iDevice)
		{
			var device = iDevice as Device;
			if (device.Name == null) {
				return;
			}
			if (!DeviceIsInList (device)) {
				_devices.Add (device);
				RunOnUiThread (() => {
					_spinnerAdapter.Add (device.Name);
					_spinnerAdapter.NotifyDataSetChanged ();
				});
			}
		}

		// <summary>
		// Check if the device is already found
		// Because the Rssi is saved in the object we can not use the contains method
		// </summary>
		// <param name="device">the device we want to check</param>
		Boolean DeviceIsInList (Device device)
		{
			foreach (Device deviceInList in _devices) {
				if (deviceInList.Name.Equals (device.Name)) {
					return true;
				}
			}
			return false;
		}

		// <summary>
		// Init an empty list
		// </summary>
		void InitEmptyLists ()
		{
			_devices.Clear ();
			_spinnerAdapter.Clear ();
			_spinnerAdapter.Add ("Kein Sensor ausgewählt");
			_spinnerAdapter.NotifyDataSetChanged ();
		}

		// <summary>
		// Connect to a device from the list
		// <summary>
		// <param name="device">the device we want to connect</param>
		void ConnectToDevice (Device device)
		{
			Toast.MakeText (this, "Connect to sensor: " + device.Name, ToastLength.Long).Show ();
			_deviceService.saveDeviceForConnection (device as IDevice);
			StartActivity (typeof(SensorControlActivity));
		}

		// <summary>
		// Called on a click on a view
		// </summary>
		// <param name="v">the view</param>
		public void OnClick (View v)
		{
			if (v.Id == Resource.Id.connect_to_device) {
				if (_nearDevices.SelectedItemPosition != 0) {
					ConnectToDevice (_devices [_nearDevices.SelectedItemPosition - 1]);
				} else {
					Toast.MakeText (this, "Kein externer Sensor ausgewählt!", ToastLength.Long).Show ();
				}
			}
		}

		// <summary>
		// Will be called when a switch changes his status
		// </summary>
		// <param name="buttonView">the button</param>
		// <param name="isChecked">if the switch is set to true or false</param>
		public void OnCheckedChanged (CompoundButton buttonView, bool isChecked)
		{
			if (buttonView.Id == Resource.Id.search_for_devices) {
				if (IsBLEAvailable ()) {
					ScanningForDevices (isChecked);
				} else {
					Toast.MakeText (this, "Bluetooth ist nicht aktiviert!", ToastLength.Long).Show ();
					buttonView.Checked = false;
				}
			}
		}
	}
}
